feat: traces IN agno

kausmeows opened the pull request:
## Summary

All things traces and spans in your own db with Agno so no data leaves your infra.

DB-
<img width="1019" height="240" alt="image" src="https://github.com/user-attachments/assets/f0e759d0-46ea-4eb6-a392-69fdf651ebd1" />

How it looks on the AgentOS platform-

You now have full observability of your agent runs along with sessions, metrics all in one platform!

<img width="935" height="705" alt="image" src="https://github.com/user-attachments/assets/273eb744-d97c-4675-bac1-39729c15824a" />

Printing traces in cli (fetching from DB)
<img width="871" height="972" alt="image" src="https://github.com/user-attachments/assets/67fc4c54-0b8d-493b-82ef-e2516c599ddc" />

Endpoints-
- `/traces`
- `/traces/<trace_id>`
- `/traces/<trace_id>?<span_id>`

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [ ] Code complies with style guidelines
- [ ] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [ ] Self-review completed
- [ ] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [ ] Tested in clean environment
- [ ] Tests added/updated (if applicable)

---

## Additional Notes

Add any important context (deployment instructions, screenshots, security considerations, etc.)

kausmeows committed:
feat: playing around

kausmeows committed:
update

kausmeows committed:
update

kausmeows committed:
chore: update cookbook

kausmeows committed:
update

kausmeows committed:
chore: update for trace and span seperate table

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
update

kausmeows committed:
update

kausmeows committed:
update

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
update

kausmeows committed:
update

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: update /traces and /traces/<trace_id> routes

kausmeows committed:
update

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
fix: mypy

kausmeows committed:
fix: tests

kausmeows committed:
chore: add db_id and pagination in traces routes

kausmeows committed:
fix: mypy

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
fix: mypy cookbook

kausmeows committed:
fix: tests

kausmeows committed:
chore: add endpoint to get span data

kausmeows committed:
chore: add traces db info in os config api

kausmeows committed:
chore: update span data route

kausmeows committed:
update

kausmeows committed:
chore: update /traces/{trace_id}?span_id

kausmeows commented on libs/agno/agno/os/router.py:
```diff
@@ -1718,4 +1726,285 @@ async def cancel_workflow_run(workflow_id: str, run_id: str):
+
+        # Verify the database has trace support
+        if not hasattr(db, "get_traces"):
+            raise HTTPException(status_code=500, detail="Selected database does not support traces")
```
yeah i had this when i just had those functions for `sqlite`, so this was needed but since now i added the functions for other dbs (not implemented ones) this is actually not needed in this way.

kausmeows commented on libs/agno/agno/os/router.py:
```diff
@@ -1718,4 +1726,285 @@ async def cancel_workflow_run(workflow_id: str, run_id: str):
+            traces, total_count = result
+
+            end_time_ms = time_module.time() * 1000
+            search_time_ms = round(end_time_ms - start_time_ms, 2)
```
This actually is a field in the class `PaginationInfo` so do we not want to have it? 

```python
class PaginationInfo(BaseModel):
    page: int = Field(0, description="Current page number (0-indexed)", ge=0)
    limit: int = Field(20, description="Number of items per page", ge=1, le=100)
    total_pages: int = Field(0, description="Total number of pages", ge=0)
    total_count: int = Field(0, description="Total count of items", ge=0)
    search_time_ms: float = Field(0, description="Search execution time in milliseconds", ge=0
)
```

kausmeows commented on libs/agno/agno/os/schema.py:
```diff
@@ -1050,3 +1059,260 @@ class PaginatedResponse(BaseModel, Generic[T]):
+    end_time: int = Field(..., description="End time in nanoseconds")
+    status: str = Field(..., description="Status code (OK, ERROR)")
+    spans: Optional[List["TraceNode"]] = Field(None, description="Child spans in the trace hierarchy")
+    metadata: Optional[Dict[str, Any]] = Field(None, description="Additional span attributes and data")
```
yeah we could use `metadata`, but rn metadata stores things like `tool_names`, `parameters` for each spans but i think we should have `extra_data` seperately not to confuse thing, will add!

manuhortet replied:
Yes good! metadata is a different thing indeed, we should mix usages

dirkbrnd commented on cookbook/integrations/observability/trace_to_database.py:
```diff
@@ -0,0 +1,218 @@
+
+try:
+    # Get all traces
+    traces = db.get_traces(limit=10, run_id=response.run_id)
```
a single run would only have 1 trace? Should it not be get_trace?

kausmeows committed:
update get_trace function

kausmeows committed:
update

kausmeows committed:
update

kausmeows committed:
update

kausmeows committed:
chore: update create_spans

kausmeows committed:
update

kausmeows committed:
chore: update extra_data field

kausmeows committed:
chore: add enable_tracing flag

kausmeows committed:
chore: enable_tracing flag on AgentOS

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: add teams agentos example

dirkbrnd commented on libs/agno/agno/db/sqlite/schemas.py:
```diff
@@ -93,6 +93,38 @@
+    "start_time_ns": {"type": BigInteger, "nullable": False, "index": True},
+    "end_time_ns": {"type": BigInteger, "nullable": False},
+    "duration_ms": {"type": BigInteger, "nullable": False},
+    "total_spans": {"type": BigInteger, "nullable": False},
```
Not sure about this one. It can be derived right? What if you add more spans? This is a weird one to have for a join relationship.

kausmeows replied:
updated to be at api level

dirkbrnd commented on libs/agno/agno/db/sqlite/schemas.py:
```diff
@@ -93,6 +93,38 @@
+
+SPAN_TABLE_SCHEMA = {
+    "span_id": {"type": String, "primary_key": True, "nullable": False},
+    "trace_id": {"type": String, "nullable": False, "index": True},
```
Can we define a FK on this?

kausmeows replied:
yes, updated

dirkbrnd commented on libs/agno/agno/db/sqlite/schemas.py:
```diff
@@ -93,6 +93,38 @@
+    "end_time_ns": {"type": BigInteger, "nullable": False},
+    "duration_ms": {"type": BigInteger, "nullable": False},
+    "attributes": {"type": JSON, "nullable": True},
+    "events": {"type": JSON, "nullable": True},
```
What are these?

dirkbrnd commented on libs/agno/agno/db/sqlite/sqlite.py:
```diff
@@ -48,6 +48,8 @@ def __init__(
         metrics_table: Optional[str] = None,
         eval_table: Optional[str] = None,
         knowledge_table: Optional[str] = None,
+        trace_table: Optional[str] = None,
```
`traces_table`, `spans_table`. We made a mistake with the others, so best to be correct going forward

dirkbrnd commented on libs/agno/agno/db/sqlite/sqlite.py:
```diff
@@ -1974,6 +1995,322 @@ def rename_eval_run(
 
+    # -- Trace methods --
+
+    def create_trace(self, trace) -> None:
```
annotation?

dirkbrnd commented on libs/agno/agno/db/sqlite/sqlite.py:
```diff
@@ -1974,6 +1995,322 @@ def rename_eval_run(
+        self,
+        trace_id: Optional[str] = None,
+        run_id: Optional[str] = None,
+        session_id: Optional[str] = None,
```
That doesn't make sense as a filter right? A trace_id would be unique? It should be a 1to1 with run_id, so that might make sense, but the others don't make sense.

kausmeows replied:
yeah ok that makes sense.

dirkbrnd commented on libs/agno/agno/db/sqlite/sqlite.py:
```diff
@@ -1974,6 +1995,322 @@ def rename_eval_run(
+                # Use INSERT OR IGNORE to avoid duplicate spans
+                for span in spans:
+                    stmt = sqlite.insert(table).values(span.to_dict())
+                    stmt = stmt.on_conflict_do_nothing(index_elements=["span_id"])
```
Do nothing? But we have a primary key?

dirkbrnd commented on libs/agno/agno/db/sqlite/sqlite.py:
```diff
@@ -1974,6 +1995,322 @@ def rename_eval_run(
+        self,
+        trace_id: Optional[str] = None,
+        parent_span_id: Optional[str] = None,
+        limit: Optional[int] = 1000,
```
limit without page?

dirkbrnd commented on libs/agno/agno/os/router.py:
```diff
@@ -1718,4 +1726,278 @@ async def cancel_workflow_run(workflow_id: str, run_id: str):
+        db = get_db(os.dbs, db_id)
+
+        try:
+            start_time_ms = time_module.time() * 1000
```
Btw check how we do time-based filers on the session router. Lets be consistent with that.

dirkbrnd commented on libs/agno/agno/tracing/setup.py:
```diff
@@ -0,0 +1,109 @@
+
+def setup_tracing(
+    db: Union[BaseDb, AsyncBaseDb],
+    use_batch_processor: bool = False,
```
```suggestion
    batch_processing: bool = False,
```

kausmeows committed:
chore: add trace session stats endpoing

kausmeows committed:
update

kausmeows committed:
update

kausmeows committed:
update

kausmeows committed:
fix: annotation

kausmeows committed:
update

kausmeows commented on libs/agno/agno/os/router.py:
```diff
@@ -1718,4 +1726,278 @@ async def cancel_workflow_run(workflow_id: str, run_id: str):
+    )
+    async def get_trace(
+        trace_id: str,
+        span_id: Optional[str] = Query(default=None, description="Optional: Span ID to retrieve specific span"),
```
its just a query param right? 
`/trace/{trace_id}?span_id="..."` basically "fetch me this span of the {trace_id}'s trace"

how else should we do this? make another span endpoint like- `/span/{span_id}`?

kausmeows committed:
update filters

kausmeows committed:
chore: input output at top level

kausmeows commented on libs/agno/agno/db/sqlite/schemas.py:
```diff
@@ -93,6 +93,38 @@
+    "start_time_ns": {"type": BigInteger, "nullable": False, "index": True},
+    "end_time_ns": {"type": BigInteger, "nullable": False},
+    "duration_ms": {"type": BigInteger, "nullable": False},
+    "attributes": {"type": JSON, "nullable": True},
```
>I think we can add input/output top-level

Do we need them as columns in db? we already do send it as root level in the api

<img width="573" height="575" alt="image" src="https://github.com/user-attachments/assets/c4b2fc93-5fb2-46bd-8c5d-76a3ef49602b" />

dirkbrnd replied:
Hmm ok yeah that is good

kausmeows commented on libs/agno/agno/os/schema.py:
```diff
@@ -1050,3 +1059,265 @@ class PaginatedResponse(BaseModel, Generic[T]):
+    start_time: int = Field(..., description="Start time in nanoseconds")
+    end_time: int = Field(..., description="End time in nanoseconds")
+    status: str = Field(..., description="Status code (OK, ERROR)")
+    spans: Optional[List["TraceNode"]] = Field(None, description="Child spans in the trace hierarchy")
```
Oh we can rename, Initially Ruan showed me a schema he was having in FE as demo data which had this `TraceNode` structure, so that just stuck in my mind.

kausmeows committed:
chore: fk update

kausmeows committed:
update

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: enable_tracing -> tracing

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: remove on conflict do nothing

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
update

kausmeows committed:
update

kausmeows commented on libs/agno/agno/os/router.py:
```diff
@@ -1718,4 +1726,278 @@ async def cancel_workflow_run(workflow_id: str, run_id: str):
         return JSONResponse(content={}, status_code=200)
 
+    # -- Trace routes ---
+    @router.get(
```
will add them as a seperate router

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: update total_spans and error_count to be at api level

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: add more examples

kausmeows committed:
chore: move traces routes in a seperate new router

kausmeows committed:
chore: fix mypy style

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: update error top level

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

dirkbrnd commented on libs/agno/agno/db/sqlite/schemas.py:
```diff
@@ -93,6 +93,38 @@
+    "run_id": {"type": String, "nullable": True, "index": True},
+    "session_id": {"type": String, "nullable": True, "index": True},
+    "user_id": {"type": String, "nullable": True, "index": True},
+    "agent_id": {"type": String, "nullable": True, "index": True},
```
How about rather having `type` and `entity_id`? Then we can support anything in the future?

kausmeows committed:
chore: add team_id for stats

kausmeows committed:
fix: mypy

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: update trace router to infer sync/async db

pritipsingh committed:
Chore: Add Tracing demo example (#5337)

## Summary

Describe key changes, mention related issues or motivation for the
changes.

(If applicable, issue number: #\_\_\_\_)

## Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [ ] Code complies with style guidelines
- [ ] Ran format/validation scripts (`./scripts/format.sh` and
`./scripts/validate.sh`)
- [ ] Self-review completed
- [ ] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included
or updated (if applicable)
- [ ] Tested in clean environment
- [ ] Tests added/updated (if applicable)

---

## Additional Notes

Add any important context (deployment instructions, screenshots,
security considerations, etc.)

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: update time to be stored as datetime

kausmeows committed:
chore: add time filter in trace stats api

kausmeows committed:
chore: store in iso read in datetime

kausmeows committed:
update

kausmeows committed:
update

kausmeows committed:
chore: time aware conversion

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: update workflow tracing and root span name logic

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: fix test

manuhortet committed:
fix: schema versions table handling on OS auto provision

kausmeows committed:
chore: update timing related changes

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: add missing import after conflict resolution

kausmeows committed:
fix: add workflow_id in /traces api

kausmeows committed:
chore: workflow run metadata

kausmeows committed:
chore: put workflow step type in its spans for node level info

kausmeows committed:
fix: mypy

kausmeows committed:
chore: step_type out of metadata

kausmeows committed:
chore: workflow output in the root run level

kausmeows committed:
fix: mypy and indentation

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
fix: sqlite table creation issue

kausmeows committed:
chore: CHAIN -> WORKFLOW, and for TEAM too

kausmeows committed:
chore: traces db parity (#5518)

## Summary

update all DBs to support traces
- [x] Sqlite (sync + async)
- [x] Postgres (sync + async)
- [x] MySQL
- [x] Mongo (sync + async)
- [x] JsonDB
- [x] GCS_Json
- [x] Firestore
- [x] DynamoDB
- [x] Redis
- [x] Singlestore
- [x] SurrealDB

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [ ] Code complies with style guidelines
- [ ] Ran format/validation scripts (`./scripts/format.sh` and
`./scripts/validate.sh`)
- [ ] Self-review completed
- [ ] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included
or updated (if applicable)
- [ ] Tested in clean environment
- [ ] Tests added/updated (if applicable)

---

## Additional Notes

Add any important context (deployment instructions, screenshots,
security considerations, etc.)

---------

Co-authored-by: Monalisha Mishra <mishramonalisha76@gmail.com>

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
update

kausmeows committed:
fix: mypy

kausmeows committed:
chore: update import without type checking in all dbs

kausmeows committed:
fix: mypy for unused import

kausmeows committed:
fix: mypy

kausmeows commented on libs/agno/agno/db/sqlite/async_sqlite.py:
```diff
@@ -27,6 +27,9 @@
 from agno.utils.string import generate_id
 
+if TYPE_CHECKING:
+    from agno.tracing.schemas import Span, Trace
```
yes, not needed. pushed an update

manuhortet committed:
Merge branch 'main' into feat/traces-in-agno

manuhortet commented on libs/agno/agno/db/postgres/postgres.py:
```diff
@@ -2319,3 +2351,497 @@ def migrate_table_from_v1_to_v2(self, v1_db_schema: str, v1_table_name: str, v1_
+            )
+        else:
+            # Fallback if spans table doesn't exist
+            return select(table, literal(0).label("total_spans"), literal(0).label("error_count"))
```
We don't need queries if there's no spans table. Let's remove this and when calling this function (in create_trace), if spans_table is None, just skip the querying logic and return None.

Not sure if it applies to more dbs

manuhortet replied:
I missed that the trace can exist without spans - I see the point of this now vs exiting earlier in the previous functions. Let's keep, all good!

manuhortet commented on libs/agno/agno/db/sqlite/schemas.py:
```diff
@@ -95,6 +95,42 @@
+    "start_time": {"type": String, "nullable": False, "index": True},  # ISO 8601 datetime string
+    "end_time": {"type": String, "nullable": False},  # ISO 8601 datetime string
+    "duration_ms": {"type": BigInteger, "nullable": False},
+    "attributes": {"type": JSON, "nullable": True},
```
We should use `JSONB` here

dirkbrnd commented on libs/agno/agno/os/routers/traces/traces.py:
```diff
@@ -0,0 +1,612 @@
+
+            trace_inputs = {}
+            for trace in traces:
+                spans_result = db.get_spans(trace_id=trace.trace_id)
```
rather check for async DB or not

dirkbrnd commented on libs/agno/tests/unit/memory/test_memory_manager_async.py:
```diff
@@ -1,11 +1,14 @@
+    from agno.tracing.schemas import Span, Trace
+
 
 class DummyAsyncMemoryDb(AsyncBaseDb):
```
This is a very weird place to define this... We should move this to a higher up file

kausmeows committed:
chore: refactor for tracing=True

kausmeows commented on cookbook/agent_os/tracing/01_basic_agent_tracing.py:
```diff
@@ -0,0 +1,37 @@
+agent_os = AgentOS(
+    description="Example app for tracing HackerNews",
+    agents=[agent],
+)
```
yes, have refactored the `tracing=True` case as we discussed on slack thread

kausmeows committed:
chore: update create_table_if_not_found in some dbs

kausmeows committed:
chore: update for sync/async db trace/span calls

kausmeows committed:
chore: add util for parsing iso datetime to UTC

kausmeows committed:
chore: util for duration ms

kausmeows commented on libs/agno/agno/os/routers/traces/schemas.py:
```diff
@@ -0,0 +1,435 @@
+        if duration_ms < 1000:
+            duration_str = f"{duration_ms}ms"
+        else:
+            duration_str = f"{duration_ms / 1000:.2f}s"
```
yes, lets do that. Updated!

kausmeows committed:
chore: update tracing flag for demo cookbook

kausmeows commented on cookbook/demo/traces/basic_agent_tracing.py:
pushed an update here- https://github.com/agno-agi/agno/pull/5155/commits/114dbc1c16e6b6044549098b9aa6ab26038f66c6

dirkbrnd replied:
@kausmeows yeah that is what I meant. We don't need new agents really

kausmeows committed:
chore: remove traces README

kausmeows committed:
chore: remove demo traces files

manuhortet committed:
Merge branch 'main' into feat/traces-in-agno

manuhortet commented on libs/agno/agno/os/app.py:
```diff
@@ -369,6 +380,37 @@ def _initialize_workflows(self) -> None:
 
+    def _setup_tracing(self) -> None:
+        """Set up OpenTelemetry tracing for this AgentOS."""
+        # Find the first available database
```
Don't we want to call setup_tracing for all available dbs?

Also if we call _setup_tracing after _auto_discover_databases we can avoid iterating over agents etc, and just read self.dbs -> that means calling this in the get_app logic instead of the init, but that's okay

manuhortet reviewed (approved):
Great work!

kausmeows committed:
chore: update for supporting custom table name

kausmeows committed:
chore: cleanup logs

kausmeows committed:
chore: add `tracing_db` support for multi-db scenario (#5574)

## Summary

- so in a case where multiple agents can have different db instances we
need to make sure to have a right DX that shows that traces are exported
to only one database for an OS. So its recommended to set its own
`tracing_db`.
- if no `tracing_db` is set, it will pickup from the first db it finds
in the OS. We will warn about this behaviour in the docs

- the `/config` already handles it for this case so when we have a
seperate tracing db, it works out of the box in FE.

<img width="1499" height="256" alt="image"
src="https://github.com/user-attachments/assets/4c1522f3-95d0-4a96-b1fc-abfc565ec0ef"
/>

## Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [x] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [ ] Code complies with style guidelines
- [ ] Ran format/validation scripts (`./scripts/format.sh` and
`./scripts/validate.sh`)
- [ ] Self-review completed
- [ ] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included
or updated (if applicable)
- [ ] Tested in clean environment
- [ ] Tests added/updated (if applicable)

---

## Additional Notes

Add any important context (deployment instructions, screenshots,
security considerations, etc.)

kausmeows committed:
chore: remove unused import

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
Merge branch 'main' into feat/traces-in-agno

kausmeows committed:
chore: cookbook update
